### DISCUSSION TITLE
Browser wallet connector

### DISCUSSION THREAD

**Discussion by askindus** [bmresearch/Solnet#212] [Q&A]
I have successfully tested Solnet and minting new NFT tokens works very well. I would now offer users on my website the possibility to buy nfts directly from their browser wallet like "Phantom". However, most frameworks are based on react. Is there a solution for the .NET world to run such a browser wallet connector in a MVC Core application?

**Comment by hoakbuilds** [comment]
I haven't really investigated from the MVC Core point of view but I've integrated the npm packages which have the wallet adapters for the browser into a Blazor Wasm dapp-scaffold project to easily bootstrap apps which need wallet adapter connection, although this is still very much broken and requires a lot of improvements because of JS Interop, someone is actively working on improving it.

You can find that here
https://github.com/bmresearch/blazor-wasm-dapp-scaffold

**Reply by Liebeck** [reply]
@murlokito  Thanks for pointing out the repository. I was able to get it running but, unfortunately, receive errors when selecting Phantom wallet. How much work do you think is necessary to use that project as production site which is able to select a wallet and transfer Lamports to another wallet?
I'm currently looking what the best way is to go for a .NET based approach with the Solana wallet adapter

**Reply by hoakbuilds** [reply]
As pointed out in the README, there are a lot of things that need to be done to have proper JS interop so it's still very much away from being actually a good skeleton to bootstrap apps for production use, this is unfortunate because I know there's demand for it but unfortunately even Solnet as a library is very recent and the core team that started Solnet has been literally all over the place which makes it very hard to move forward with this kind of things.

I am not very experienced with Blazor Wasm myself which is why I only bootstrapped the scaffold project and left it like that because I had to focus elsewhere, I know someone is working on improving it but I don't know what progress has been made since we last talked. But I reckon someone with good Blazor Wasm knowledge could move fairly quickly with it now that it has been bootstrapped as is.

As mentioned, I think the things that need to be done are:
- writing a class in the JsLib which abstracts the wallet adapters and provides ability to plug into it event handlers with `JSInvokable` attribute, because otherwise you can't listen to the events emitted by the underlying `EventEmitter`
- that same class should abstract the `SignTransaction(byte[])` method and reconstruct the `Transaction` object in js before actually requesting a signature, because `SignMessage` is used to sign arbitrary messages and only Phantom exposes it

**Reply by martinwhitman** [reply]
What errors are you getting when selecting Phantom? I wasn't getting any errors at that point. I am trying today to finish connecting the event emitter from the adapter to the event listener in .NET that I think I have set up right.

**Reply by martinwhitman** [reply]
@Liebeck Have you been able to get this to work? PS the PR I submitted contains the improvements murlokito suggested.

**Comment by martinwhitman** [comment]
@askinadze Are you looking to build a dApp? To my mind Blazor is possibly the perfect dotnet option for that, and you can combine that with Core API controllers. Not that a MVC Core scaffold would not be useful.
